### PR TITLE
Implement CRUD operations for Application Templates

### PR DESCRIPTION
**Pull request by kfurgol**
**Description**

Changes proposed in this pull request:

- Implemented resolvers and services for CRUD operations within Application Templates domain
- Added unit tests
<details>
<summary>Helpers for review</summary>
<p>

```
query applicationTemplate{
  applicationTemplate(id: "") {
    id
    name
    description
    applicationInput
    placeholders {
      name
      description
    }
    accessLevel
  }
}
query applicationTemplates{
  applicationTemplates {
    data {
      id
      name
      description
      applicationInput
      placeholders {
        name
        description
      }
      accessLevel
    }
  }
}

mutation createApplicationTemplate {
  createApplicationTemplate(in: {
    name: "second-template"
    description: "first-template-description"
    applicationInput: {
      name: "app"
    	description: "app-desc"
    }
    placeholders: {
      name: "first-placeholder"
      description: "first-placeholder-desc"
    }
    accessLevel: GLOBAL
  }) {
     id
    name
    description
    applicationInput
    placeholders {
      name
      description
    }
    accessLevel
  }
}

mutation updateApplicationTemplate {
  updateApplicationTemplate(id:"", in: {
    name: "updated-first-template"
    applicationInput: {
      name: "updated-app"
    }
    accessLevel: GLOBAL
  }){ 
  id
    name
  }
}

mutation deleteApplicationTemplate {
  deleteApplicationTemplate(id:""){
    id
    name
    description
    applicationInput
    placeholders {
      name
      description
    }
    accessLevel
  }
}
```

</p></details>

**Related issue(s)**
#434

### CODE REVIEW

File: components/director/internal/domain/apptemplate/resolver.go [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+func (r *Resolver) ApplicationTemplate(ctx context.Context, id string) (*graphql.ApplicationTemplate, error) {
+	tx, err := r.transact.Begin()
+	if err != nil {
+		return nil, err
+	}
+	defer r.transact.RollbackUnlessCommited(tx)
+
+	ctx = persistence.SaveToContext(ctx, tx)
+
+	is, err := r.appTemplateSvc.Get(ctx, id)
```
Comment (tgorgol) [inline components/director/internal/domain/apptemplate/resolver.go]: `is`?

File: components/director/internal/domain/apptemplate/service.go [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+	err := s.appTemplateRepo.Create(ctx, appTemplate)
+	if err != nil {
+		return "", errors.Wrap(err, "while creating Application Template")
+	}
+
+	return id, nil
+}
+
+func (s *service) Get(ctx context.Context, id string) (*model.ApplicationTemplate, error) {
+	intSys, err := s.appTemplateRepo.Get(ctx, id)
```
Comment (tgorgol) [inline components/director/internal/domain/apptemplate/service.go]: `intSys`

File: components/director/internal/domain/apptemplate/service_test.go [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+				appTemplateRepo := &automock.ApplicationTemplateRepository{}
+				appTemplateRepo.On("Create", ctx, *modelAppTemplate).Return(nil).Once()
+				return appTemplateRepo
+			},
+			ExpectedOutput: testID,
+		},
+		{
+			Name: "Error when creating",
+			AppTemplateRepoFn: func() *automock.ApplicationTemplateRepository {
+				intSysRepo := &automock.ApplicationTemplateRepository{}
```
Comment (tgorgol) [inline components/director/internal/domain/apptemplate/service_test.go]: `intSysRepo`

File: components/director/internal/domain/apptemplate/service_test.go [hunk 4]
```diff
@@ -0,0 +1,364 @@
...
+			Name: "Success",
+			AppTemplateRepoFn: func() *automock.ApplicationTemplateRepository {
+				appTemplateRepo := &automock.ApplicationTemplateRepository{}
+				appTemplateRepo.On("Get", ctx, testID).Return(modelAppTemplate, nil).Once()
+				return appTemplateRepo
+			},
+			ExpectedOutput: modelAppTemplate,
+		},
+		{
+			Name: "Error when getting integration system",
```
Comment (tgorgol) [inline components/director/internal/domain/apptemplate/service_test.go]: `integration system`

File: components/director/internal/domain/apptemplate/service_test.go [hunk 5]
```diff
@@ -0,0 +1,364 @@
...
+			Name: "Success",
+			AppTemplateRepoFn: func() *automock.ApplicationTemplateRepository {
+				appTemplateRepo := &automock.ApplicationTemplateRepository{}
+				appTemplateRepo.On("Exists", ctx, testID).Return(true, nil).Once()
+				return appTemplateRepo
+			},
+			ExpectedOutput: true,
+		},
+		{
+			Name: "Error when getting integration system",
```
Comment (tgorgol) [inline components/director/internal/domain/apptemplate/service_test.go]: `integration system` (few more occurances in tests, please check)

**Comment by crabtree** [comment]
> When I create Application, I am using the following syntax:
> 
> ```
> mutation {
>   createApplication(in:{integrationSystemID:"xxx", name:"yyy"}) {
>     id
>     name
>   }
> }
> ```
> 
> but template returns input in the following form:
> 
> ```
>       "applicationInput": "{\"Name\":\"my app\",\"Description\":null,\"Labels\":null,\"HealthCheckURL\":null,\"Webhooks\":null,\"Apis\":null,\"EventAPIs\":null,\"Documents\":null,\"IntegrationSystemID\":null}"
>         }
> ```
> 
> notice: `name` vs `Name`. If you generate JSON, you should not use it from the model, rather from Graphql object that has defined JSON tags.
> 
> There is another difference between GQL and JSON, param names (like `integrationSystemID`, `name`) are not enclosed into quotes, because of that, in tests we have:
> 
> ```
> // graphqlizer is responsible for converting Go objects to input arguments in graphql format
> type graphqlizer struct{}
> ```
> 
> for generating examples.
> We have 2 options:
> 
> * define "graphqlizer" for ApplicationINput, but how do we ensure that after adding new field to type we update it?
> * mark explicitly that appplicationINput is in JSON format (probably easier)
> 
> @crabtree what do you think?

For now, and only for some time, I would go with defining a new graphqlizer. This should be a temporary solution till the time we implement proper way of serializing the structures to qraphql. We should investigate whether there is already a proper solution similar to XML and JSON serialization. If there is not, then we should work on our own implementation. 

Regarding the implementation, this may be a good reference to start with: https://blog.gopheracademy.com/advent-2017/custom-json-unmarshaler-for-graphql-client/

File: components/director/internal/domain/apptemplate/resolver.go (hunk 1 above)
Comment (kfurgol) [reply components/director/internal/domain/apptemplate/resolver.go]: refactored

File: components/director/internal/domain/apptemplate/service_test.go [hunk 6]
```diff
@@ -0,0 +1,408 @@
...
+				},
+			},
+			AppTemplateRepoFn: func() *automock.ApplicationTemplateRepository {
+				appTemplateRepo := &automock.ApplicationTemplateRepository{}
+				return appTemplateRepo
+			},
+			ExpectedError: errors.New("while validating uniqueness of placeholders"),
+		},
+		{
+			Name:  "Error when creating",
```
Comment (aszecowka) [inline components/director/internal/domain/apptemplate/service_test.go]: name of this test will be:
`TestService_Create/Error_when_creating`
not good, not terrible

File: components/director/internal/domain/apptemplate/converter.go [hunk 7]
```diff
@@ -3,31 +3,96 @@ package apptemplate
...
 	if in == nil {
 		return nil, nil
 	}
 
-	placeholders, err := c.packPlaceholders(in.Placeholders)
-	if err != nil {
-		return nil, errors.Wrap(err, "while packing Placeholders")
+	var gqlAppInput string
+	var err error
+	if in.ApplicationInput != "" {
```
Comment (aszecowka) [inline components/director/internal/domain/apptemplate/converter.go]: if ` in.ApplicationInput` == "" then we should return error.

File: components/director/internal/domain/apptemplate/converter.go [hunk 8]
```diff
@@ -52,21 +117,33 @@ func (c *converter) FromEntity(entity *Entity) (*model.ApplicationTemplate, erro
...
 		Name:             entity.Name,
 		Description:      repo.StringPtrFromNullableString(entity.Description),
-		ApplicationInput: appInput,
+		ApplicationInput: entity.ApplicationInput,
 		Placeholders:     placeholders,
 		AccessLevel:      model.ApplicationTemplateAccessLevel(entity.AccessLevel),
 	}, nil
 }
 
+func graphqliseApplicationCreateInput(applicationInput string) (string, error) {
```
Comment (aszecowka) [inline components/director/internal/domain/apptemplate/converter.go]: make this function converter's method

File: components/director/internal/domain/apptemplate/converter.go [hunk 9]
```diff
@@ -81,7 +158,7 @@ func (c *converter) unpackApplicationInput(in string) (*model.ApplicationCreateI
 	return &appInput, nil
 }
 
-func (c *converter) packApplicationInput(in *model.ApplicationCreateInput) (string, error) {
+func (c *converter) packModelApplicationInput(in *model.ApplicationCreateInput) (string, error) {
```
Comment (aszecowka) [inline components/director/internal/domain/apptemplate/converter.go]: ApplicationCreateInputToJSON ? the same for unpack.
Also, if we store applictionCreateInput in string as a JSON, we should mark that it is storing json by appending `JSON` to the field name. The same for Entity.Placeholders

File: components/director/internal/domain/apptemplate/service.go [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+	}
+}
+
+func (s *service) Create(ctx context.Context, in model.ApplicationTemplateInput) (string, error) {
+	id := s.uidService.Generate()
+	appTemplate := in.ToApplicationTemplate(id)
+
+	ok := checkIfPlaceholdersAreUnique(appTemplate.Placeholders)
+	if !ok {
+		return "", errors.New("while validating uniqueness of placeholders")
```
[the remark this answers is not in the document]
Comment (kfurgol) [reply components/director/internal/domain/apptemplate/service.go]: refactored, I've followed the convention used in labeldef domain `[name=xyz]`

**Pull request merged**